# Kapitel 5 Dimensionsreducering

## Uppgift 1
Curse of dimensionality (dimensionalitetens förbannelse) innebär att när antalet dimensioner ökar växer rymden så mycket att datapunkterna hamnar väldigt glest. Det gör det svårt för modeller att hitta mönster och ökar risken för överanpassning.

## Uppgift 2
Dimensionsreducering innebär att minska antalet variabler och samtidigt behålla så mycket viktig information som möjligt.
Det görs för att snabba upp beräkningar, minska minnesanvändning, minska risken för överanpassning och göra det möjligt att visualisera datan i 2D eller 3D.

## Uppgift 3
Hur PCA fungerar (figur 5.4, s. 224):
PCA hittar de riktningar (principalkomponenter) där datan har störst spridning/varians. Den första komponenten läggs i riktningen med mest variation, och nästa komponent vinkelrätt mot den första för att fånga näst mest variation. Därefter projiceras datapunkterna ner på dessa komponenter.

## Uppgift 5
Stina har fel i att maximal prediktion alltid är det enda som räknas, och Kalle har rätt i att tid och resurser är viktiga. I verkligheten måste en modell ofta leverera snabba svar och vara resurssnål. En något enklare modell som är snabb och billig att köra är ofta bättre i praktiken.

## Uppgift 6
Efter en PCA motsvarar de nya variablerna (principalkomponenterna) linjära kombinationer av alla ursprungliga variabler. Detta gör att den direkta tolkningen av enskilda variabler (som t.ex. hästkrafter eller ålder) försvinner.

## Uppgift 8
Förklara vad nedanstående kod gör.

In [1]:
import numpy as np
from sklearn.decomposition import PCA

# Creating a dataset with 3 features/columns
X = np.random.rand(1000, 3)
print(X[0:5])

# Reducing the data to 2 dimensions
pca = PCA(n_components=2)
X2D = pca.fit_transform(X)
print(X2D[0:5])

# 'Recreating' the data to 3 dimensions
X3D_inv = pca.inverse_transform(X2D)

# Not exactly equal since some information was lost in the transformation
print(np.allclose(X3D_inv, X))

[[0.3339808  0.24143671 0.98513744]
 [0.47375421 0.05663508 0.71594811]
 [0.45275594 0.01838145 0.82821296]
 [0.58852361 0.59035378 0.76218624]
 [0.58776712 0.35272942 0.43112358]]
[[-0.0844511  -0.3905038 ]
 [ 0.19654685 -0.40161349]
 [ 0.17871128 -0.4656868 ]
 [-0.05782001  0.07628387]
 [ 0.1643679  -0.04696113]]
False


Koden skapar ett dataset med 1000 rader och 3 dimensioner, minskar det till 2 dimensioner med PCA och projicerar sedan tillbaka det till 3 dimensioner med inverse_transform. np.allclose ger False eftersom information förlorades när den tredje dimensionen togs bort.

## Uppgift 9
Genomför en PCA på car_price_dataset.csv innan modellering med ML.

In [2]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.decomposition import PCA
from sklearn.linear_model import LinearRegression
from sklearn.metrics import root_mean_squared_error

df = pd.read_csv('../kapitel3/car_price_dataset.csv', sep=';').dropna()
X = pd.get_dummies(df.drop(columns='Price'), drop_first=True)
y = df['Price']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.20, random_state=42)

# 1. Linjär regression utan PCA
model_no_pca = LinearRegression().fit(X_train, y_train)
rmse_no_pca = root_mean_squared_error(y_test, model_no_pca.predict(X_test))
print(f"RMSE utan PCA: {rmse_no_pca:.2f}")

# 2. Linjär regression med PCA (95% varians)
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

pca = PCA(n_components=0.95)
X_train_pca = pca.fit_transform(X_train_scaled)
X_test_pca = pca.transform(X_test_scaled)

model_pca = LinearRegression().fit(X_train_pca, y_train)
rmse_pca = root_mean_squared_error(y_test, model_pca.predict(X_test_pca))
print(f"RMSE med PCA: {rmse_pca:.2f}")

RMSE utan PCA: 64.91
RMSE med PCA: 870.75


Utan PCA blev RMSE 64.91 och med PCA blev den 870.75. I det här fallet blev modellen alltså mycket sämre efter PCA, eftersom oövervakad PCA tog bort detaljer och dummyvariabler som behövdes för att förutsäga bilpriset.